hudi-common/src/main/java/org/apache/hudi/avro/AvroSchemaUtils.java

```diff
@@ -44,31 +61,65 @@ public static String getAvroRecordQualifiedName(String tableName) {
+   *   <li>Schemas A and B are array schemas and element-type of B is a strict projection
+   *   of the element-type of A, or</li>
+   *   <li>Schemas A and B are map schemas and value-type of B is a strict projection
+   *   of the value-type of A, or</li>
+   *   <li>Schemas A and B are union schemas (of the same size) and every element-type of B
+   *   is a strict projection of the corresponding element-type of A, or</li>
    *   <li>Schemas A and B are record schemas and every field of the record B has corresponding
    *   counterpart (w/ the same name) in the schema A, such that the schema of the field of the schema
-   *   B is also a projection of the A field's schema</li>
+   *   B is also a strict projection of the A field's schema</li>
    * </ol>
    */
-  public static boolean isProjectionOf(Schema sourceSchema, Schema targetSchema) {
-    if (sourceSchema.getType() != Schema.Type.RECORD
-        || targetSchema.getType() != Schema.Type.RECORD) {
-      return Objects.equals(sourceSchema, targetSchema);
-    }
+  public static boolean isStrictProjectionOf(Schema sourceSchema, Schema targetSchema) {
+    return isProjectionOfInternal(sourceSchema, targetSchema, Objects::equals);
+  }
 
-    for (Schema.Field targetField : targetSchema.getFields()) {
-      Schema.Field sourceField = sourceSchema.getField(targetField.name());
-      if (sourceField == null || !isProjectionOf(sourceField.schema(), targetField.schema())) {
-        return false;
+  private static boolean isProjectionOfInternal(Schema sourceSchema,
+                                                Schema targetSchema,
+                                                BiFunction<Schema, Schema, Boolean> atomicTypeEqualityPredicate) {
+    if (sourceSchema.getType() == targetSchema.getType()) {
+      if (sourceSchema.getType() == Schema.Type.RECORD) {
+        for (Schema.Field targetField : targetSchema.getFields()) {
+          Schema.Field sourceField = sourceSchema.getField(targetField.name());
+          if (sourceField == null || !isProjectionOfInternal(sourceField.schema(), targetField.schema(), atomicTypeEqualityPredicate)) {
+            return false;
+          }
+        }
+        return true;
+      } else if (sourceSchema.getType() == Schema.Type.ARRAY) {
+        return isProjectionOfInternal(sourceSchema.getElementType(), targetSchema.getElementType(), atomicTypeEqualityPredicate);
+      } else if (sourceSchema.getType() == Schema.Type.MAP) {
+        return isProjectionOfInternal(sourceSchema.getValueType(), targetSchema.getValueType(), atomicTypeEqualityPredicate);
+      } else if (sourceSchema.getType() == Schema.Type.UNION) {
+        List<Schema> sourceNestedSchemas = sourceSchema.getTypes();
+        List<Schema> targetNestedSchemas = targetSchema.getTypes();
+        if (sourceNestedSchemas.size() != targetNestedSchemas.size()) {
+          return false;
+        }
+
+        for (int i = 0; i < sourceNestedSchemas.size(); ++i) {
```

Comment:
so, ordering of UNION also matters in equality/projection/compatibility is it ?

Reply:
Yes. We can relax it in the future if there would be a use-case for it